### ISSUE TITLE
preseed buster removes python2 and leaves the vm without /usr/bin/python

### ISSUE THREAD

**Issue by chymian** [debops/debops#1079]
using the debops-preseed on buster somehow removes python2 and therefore the link from 
`/usr/bin/python -> /usr/bin/python2' and leaves the image without a python-binary.

manually installing with the same iso: "debian-10.0.0-amd64-netinstal.iso" and the same selections during install, outputs a vm with python2 and a accessible binary `/usr/bin/python`

**Comment by drybjed** [comment]
I'm not sure why preseed would remove `python2` specifically, maybe some dependency replaces it. But, currently `debops.preseed` role is very outdated, I want to update it to be able to install servers unattended again so I'll try to improve it before DebOps `v1.2.0` is released around November. Stay tuned.

**Comment by chymian** [comment]
great.
you might consider to include a "install qemu-guest-agent" (or other hypervisors's guest-agents) switch. 
hashicorps (proxmox-)packer does need to get the guest-IP from the agent after initial reboot, to successfully end the run.

**Comment by drybjed** [comment]
@chymian Sounds good, if I find a good way to detect the virtualization type with `bash` since that's the environment the postinst script runs in.

**Comment by chymian** [comment]
@drybjed, at the beginning of dmesg, there a few entries, at least for KVM-Hypervisor. I don't have XEN or VMware available to test…

```
dmesg | head -n 200 | grep -i kvm
[    0.000000] Hypervisor detected: KVM
[    0.000000] kvm-clock: Using msrs 4b564d01 and 4b564d00
[    0.000001] kvm-clock: cpu 0, msr 5e3e4001, primary cpu clock
[    0.000001] kvm-clock: using sched offset of 12081620546 cycles
[    0.000004] clocksource: kvm-clock: mask: 0xffffffffffffffff max_cycles: 0x1cd42e4dffb, max_idle_ns: 881590591483 ns
[    0.082570] Booting paravirtualized kernel on KVM
[    0.253274] KVM setup async PF for cpu 0
[    0.253284] kvm-stealtime: cpu 0, msr 5ca161c0
[    0.391938] smpboot: CPU0: Intel Common KVM processor (family: 0xf, model: 0x6, stepping: 0x1)
[    0.033140] kvm-clock: cpu 1, msr 5e3e4041, secondary cpu clock
[    0.402178] KVM setup async PF for cpu 1
[    0.402178] kvm-stealtime: cpu 1, msr 5cb161c0
```

**Comment by jankowa** [comment]
I've the same problem with an host with debian buster. No idea if there is in relation with preseeding but my host is also installed with an autoinstaller with preseeding configuration (not with debops).

Basically  I installed a base system with python2.7 and python3. While trying to bootstrap it removes all python 2.7 and doesn't find any /usr/bin/python.

Why python2.7 is purged anyway?

**Comment by drybjed** [comment]
@jankowa The `debops.python` role by default purges the Python 2.7 environment on Debian Buster to make way for Python3-only setup. This [can be disabled](https://docs.debops.org/en/master/ansible/roles/debops.python/defaults/main.html#python-2-package-purging) if you still need Py2 environment, just set the `python__raw_purge_v2` variable to `False` in the inventory.

**Comment by jankowa** [comment]
@drybjed Thanks. I'll look at this.
hm, maybe not directly related to this issue but related with python environment:
A minimal bootstrapped Debian-System (in my case an lxc-container from buster template) comes without and /usr/bin/python binary at all. Even if python3 is installed, it's only available under /usr/bin/python3. So I get:

"/bin/sh: 1: /usr/bin/python: not found\r\n",

ah, I see, this is: #1002

**Comment by jankowa** [comment]
ah, I see, this is: #1002

**Comment by drybjed** [comment]
You might also check out the [Debian Python Policy](https://www.debian.org/doc/packaging-manuals/python-policy/python.html#interpreter), it discourages use of `/usr/bin/python` in scripts - they should explicitly specify the Python version they intend to work with.

I don't plan to implement a workaround in DebOps to symlink `/usr/bin/python3` to `/usr/bin/python` because this might break things in the future.

**Comment by bleuchtang** [comment]
It's not really the same issue but I think it's related; I encounter an error with acme-tiny installed with debops-pki. On a fresh minimal debian buster acme-tiny gave me this error (located in /etc/pki/realms/domain/acme/error.log):
/usr/bin/env: 'python': No such file or directory

My workaround for now is to set python__v2 to True.

**Comment by imrejonk** [comment]
I just encountered the same problem as @bleuchtang. I'd say this makes the `pki` role currently unsuitable for managing ACME certificates on Python 3-only systems.

Reading the [Debian Python Policy](https://www.debian.org/doc/packaging-manuals/python-policy/python.html#interpreter) @drybjed linked to, I think that the best solution would be if the Python interpreter used by acme-tiny would change to `python3` instead of (the discouraged) `python`. acme-tiny seems to work fine with the Python 3 interpreter.

I opened a pull request: https://github.com/diafygi/acme-tiny/pull/240
